<a href="https://colab.research.google.com/github/Pravallikayadav2006/hushmap-project/blob/main/Internship_Agentic_AI_Patient_Health_Record_Raw.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Patient Health Records Cleaning and Analysis

## Project Objective

Healthcare datasets often contain missing values, duplicate records, inconsistent text formats, invalid entries, and outliers.

The objective of this project is to clean the patient health records dataset, improve data quality, perform exploratory data analysis, and generate meaningful insights from the cleaned data.

The cleaned dataset can support better healthcare reporting and decision-making.

https://www.kaggle.com/datasets/mursaleen880/patient-health-record-raw-csv/data :link to download the dataset

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Pandas is used for data manipulation and cleaning.
# NumPy is used for numerical operations.
# Matplotlib and Seaborn are used for visualization.

In [ ]:
from google.colab import files

uploaded = files.upload()

Saving Patient_Health_Records_Raw.csv to Patient_Health_Records_Raw.csv


In [ ]:
df = pd.read_csv('Patient_Health_Records_Raw.csv')

df.head()

,Patient_ID,Name,Age,Gender,City,BMI,Blood_Pressure,Heart_Rate,Cholesterol_Level,Diabetic,Smoker,Medications,Last_Visit_Date,Follow_Up,Diagnosis_Code,Notes,Has_Disease
0,ID-00001,"SMITH, ALICE",250,Unknown,Boston,22kg/m2,120 over 80,83,190,Unknown,yes,NaN,Apr 15 2021,30,A00,Patient feeling fine 🙂,0
1,ID-00002,"SMITH, ALICE",-5,FEMALE,newyork,25.8,NaN,500,normal,N,No,NaN,2021-07-10,NaN,unknown,"BP high, needs recheck",unknown
2,ID-00003,John Doe,-5,NaN,Nwe Yrok,22kg/m2,120 - 80,500,190,no,Former,"aspirin,metformin",Apr 06 2021,30,B20,NaN,unknown
3,ID-00004,michael brown,80,F,New York,33.4,NaN,eighty,250,y,EX-smoker,NaN,20241113,30,NaN,Follow up soon,unknown
4,ID-00005,Jane Doe,250,male,la,30.1,120 over 80,0,normal,no,No,NaN,2025-09-27,NaN,I10,Patient feeling fine 🙂,1


## Initial Dataset Exploration

Before cleaning the dataset, it is important to understand its structure, size, column names, data types, and potential data quality issues.

This helps identify missing values, incorrect formats, duplicates, and outliers that require cleaning.

In [ ]:
# Display number of rows and columns
print("Dataset Shape:", df.shape)

Dataset Shape: (10000, 17)


In [ ]:
# Display all column names
print(df.columns.tolist())

['Patient_ID', 'Name', 'Age', 'Gender', 'City', 'BMI', 'Blood_Pressure', 'Heart_Rate', 'Cholesterol_Level', 'Diabetic', 'Smoker', 'Medications', 'Last_Visit_Date', 'Follow_Up', 'Diagnosis_Code', 'Notes', 'Has_Disease']


In [ ]:
# Display data types and non-null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Patient_ID         9493 non-null   object
 1   Name               10000 non-null  object
 2   Age                10000 non-null  object
 3   Gender             8293 non-null   object
 4   City               8997 non-null   object
 5   BMI                7582 non-null   object
 6   Blood_Pressure     7970 non-null   object
 7   Heart_Rate         10000 non-null  object
 8   Cholesterol_Level  7948 non-null   object
 9   Diabetic           8359 non-null   object
 10  Smoker             6766 non-null   object
 11  Medications        5976 non-null   object
 12  Last_Visit_Date    10000 non-null  object
 13  Follow_Up          7476 non-null   object
 14  Diagnosis_Code     8325 non-null   object
 15  Notes              5976 non-null   object
 16  Has_Disease        7480 non-null   object

In [ ]:
# Summary statistics for numerical columns
df.describe(include='all')

,Patient_ID,Name,Age,Gender,City,BMI,Blood_Pressure,Heart_Rate,Cholesterol_Level,Diabetic,Smoker,Medications,Last_Visit_Date,Follow_Up,Diagnosis_Code,Notes,Has_Disease
count,9493,10000,10000,8293,8997,7582,7970,10000,7948,8359,6766,5976,10000,7476,8325,5976,7480
unique,9493,5,104,5,9,202,1431,74,4,5,4,3,5816,3,5,3,3
top,ID-10000,"LEE, Sara",-5,Unknown,Chicago,22.5,120 over 80,500,190,no,EX-smoker,"aspirin,metformin",Dec 13 2020,two weeks,B20,Follow up soon,0
freq,1,2038,2537,1719,1048,2610,2066,2515,2048,1721,1739,2011,7,2540,1700,2023,2513


In [ ]:
(df.isnull().sum() / len(df) * 100).round(2)

,0
Patient_ID,5.07
Name,0.00
Age,0.00
Gender,17.07
City,10.03
BMI,24.18
Blood_Pressure,20.30
Heart_Rate,0.00
Cholesterol_Level,20.52
Diabetic,16.41


# Creating a Working Copy
#Instead of modifying the original dataset directly, a separate copy is created.
#This preserves the raw dataset and allows data cleaning operations to be performed safely without affecting the original records.

In [ ]:
# Creating a copy preserves the original dataset.
# All cleaning operations will be performed on clean_df.
clean_df = df.copy()
print("Copy Created Successfully")

Copy Created Successfully


## Missing Values Analysis

Missing values can affect statistical analysis and visualization.

Identifying missing values is the first step before deciding how they should be handled.

In [ ]:
clean_df.isnull().sum()

,0
Patient_ID,507
Name,0
Age,0
Gender,1707
City,1003
BMI,2418
Blood_Pressure,2030
Heart_Rate,0
Cholesterol_Level,2052
Diabetic,1641


## Duplicate Record Detection

Duplicate records may lead to inaccurate statistics and repeated patient information.

Therefore, duplicate records are identified before cleaning.

In [ ]:
duplicates = clean_df.duplicated().sum()

print("Duplicate Records:", duplicates)

Duplicate Records: 0


In [ ]:
# Before cleaning, I examined unique values to identify inconsistencies and formatting issues.
print("Gender Values:")
print(clean_df['Gender'].unique())

print("\nCity Values:")
print(clean_df['City'].unique())

print("\nDiabetic Values:")
print(clean_df['Diabetic'].unique())

print("\nSmoker Values:")
print(clean_df['Smoker'].unique())

print("\n Age Values:")
print(clean_df['Age'].unique())

Gender Values:
['Unknown' 'FEMALE' nan 'F' 'male' 'M']

City Values:
['Boston' 'newyork' 'Nwe Yrok' 'New York' 'la' nan 'Chicago' 'Los Angeles'
 'NY' 'chicago']

Diabetic Values:
['Unknown' 'N' 'no' 'y' nan 'Yes']

Smoker Values:
['yes' 'No' 'Former' 'EX-smoker' nan]

 Age Values:
['250' '-5' '80' '98' 'twenty' '6' '18' '59' '46' '52' '64' '9' '100' '54'
 '85' '15' '24' '0' '21' '83' '3' '45' '48' '67' '25' '42' '99' '56' '75'
 '65' '53' '32' '13' '57' '4' '93' '62' '41' '14' '89' '79' '86' '31' '96'
 '68' '87' '33' '40' '39' '35' '12' '55' '11' '1' '43' '7' '72' '28' '84'
 '50' '27' '97' '58' '81' '78' '61' '29' '90' '44' '37' '2' '63' '66' '51'
 '10' '82' '74' '30' '19' '23' '92' '71' '5' '17' '91' '76' '47' '26' '20'
 '49' '88' '38' '69' '73' '22' '36' '95' '77' '70' '60' '94' '8' '16' '34']


The BMI column contains a mixture of numeric values, values with units, and missing records. These inconsistencies must be resolved before performing statistical analysis.

In [ ]:
clean_df['BMI'].value_counts().head(20)

,count
BMI,
22.5,2610
22kg/m2,2467
19.2,23
20.2,22
17.7,22
18.5,21
24.8,21
27.8,20
34.2,20


Blood pressure values appear in multiple formats, including standard notation, textual representations, and missing values. Standardization is required for accurate medical analysis.

In [ ]:
clean_df['Blood_Pressure'].value_counts().head(20)

,count
Blood_Pressure,
120 over 80,2066
120/80,1987
120 - 80,1957
129/77,4
118/64,4
110/75,4
130/82,4
92/85,4
115/67,4


In [ ]:
# Check age values
# Identify incorrect age entries
clean_df['Age'].value_counts().head(20)

In [ ]:
# Convert text age into numeric age
clean_df['Age'] = clean_df['Age'].replace('twenty', 20)

In [ ]:
# Convert Age column into numeric format
clean_df['Age'] = pd.to_numeric( clean_df['Age'], errors='coerce' )

In [ ]:
# Find age values less than 0 or greater than 120
invalid_age = clean_df[ (clean_df['Age'] < 0) | (clean_df['Age'] > 120) ]
print("Invalid Age Records:", len(invalid_age))

In [ ]:
# Replace invalid ages with missing values
clean_df.loc[ (clean_df['Age'] < 0) | (clean_df['Age'] > 120), 'Age' ] = pd.NA

In [ ]:
# Fill missing ages using median value
clean_df['Age'].fillna( clean_df['Age'].median(), inplace=True )

In [ ]:
clean_df['Age'] = clean_df['Age'].fillna(clean_df['Age'].median())

In [ ]:
# Verify invalid age values

invalid_age = clean_df[
    (clean_df['Age'] < 0) |
    (clean_df['Age'] > 120)
]

print("Invalid Age Records:", len(invalid_age))

In [ ]:
# Check missing values in Age after cleaning

clean_df['Age'].isnull().sum()

Before cleaning the BMI column, the values were examined to identify formatting issues, missing values, and invalid entries.

The BMI column contained:

Numeric values
Decimal values
Values with units such as "kg/m2"
Missing values

Therefore, cleaning was required before analysis.

In [ ]:
# Check BMI values-This helps identify inconsistent BMI values such as text, units, and missing entries.
clean_df['BMI'].value_counts().head(20)

Some BMI records contained units along with the value. Since BMI should be stored as a numeric value, the units were removed to ensure consistency throughout the dataset.

Example:

22kg/m2 → 22

In [ ]:
# Remove BMI units
clean_df['BMI'] = clean_df['BMI'].str.replace('kg/m2', '', regex=False)

The BMI column was converted into numeric format so that statistical analysis and visualizations could be performed.

Any non-numeric values were automatically converted into missing values (NaN).

In [ ]:
# Convert BMI to numbers
clean_df['BMI'] = pd.to_numeric(clean_df['BMI'], errors='coerce')

A valid BMI is generally expected to fall within the range of 10 to 60.

The dataset was checked for values outside this range to identify potential outliers or data entry errors.

No BMI values were found outside the acceptable range.

Therefore, no records required correction due to invalid BMI values.

In [ ]:
# Find invalid BMI values
invalid_bmi = clean_df[
    (clean_df['BMI'] < 10) |
    (clean_df['BMI'] > 60)
]

print("Invalid BMI Records:", len(invalid_bmi))

In [ ]:
# Check missing values in BMI column before cleaning

df['BMI'].isnull().sum()

Although no invalid BMI values were found, some records contained missing BMI information.

Instead of removing these records, the missing values were filled using the median BMI value.

The median was chosen because it is less affected by extreme values and provides a reliable estimate for missing data.

In [ ]:
# Fill missing BMI values
clean_df['BMI'].fillna(
    clean_df['BMI'].median(),
    inplace=True
)

After cleaning, the BMI column was verified to ensure that:

All values were numeric
Missing values were handled
No invalid BMI values remained

The column was now ready for further analysis and visualization.

In [ ]:
# Check BMI statistics
clean_df['BMI'].describe()

In [ ]:
clean_df['BMI'].isnull().sum()

The Heart Rate column was examined to identify invalid values, text entries, and unrealistic heart rate measurements.

In [ ]:
# Check heart rate values
clean_df['Heart_Rate'].value_counts().head(20)

Some records contained heart rate values in text format. These values were converted into numeric form to ensure consistency.

In [ ]:
# Convert text heart rate
clean_df['Heart_Rate'] = clean_df['Heart_Rate'].replace('eighty', 80)

The Heart Rate column was converted into numeric format to support statistical analysis and outlier detection.

In [ ]:
# Convert heart rate to numeric
clean_df['Heart_Rate'] = pd.to_numeric(clean_df['Heart_Rate'], errors='coerce')

For this project:

Less than 40 = Invalid and
Greater than 200 = Invalid

In [ ]:
# Find invalid heart rate values
# Heart rate should be between 40 and 200
invalid_hr = clean_df[
    (clean_df['Heart_Rate'] < 40) |
    (clean_df['Heart_Rate'] > 200)
]
print("Invalid Heart Rate Records:", len(invalid_hr))

Heart rate values below 40 or above 200 were considered unrealistic and were replaced with missing values for further processing

In [ ]:
# Replace invalid heart rate values
clean_df.loc[
    (clean_df['Heart_Rate'] < 40) |
    (clean_df['Heart_Rate'] > 200),
    'Heart_Rate'
] = np.nan

Instead of deleting records, missing and invalid heart rate values were replaced using the median heart rate. This helps preserve valuable patient information while maintaining data quality.

In [ ]:
# Fill missing heart rate values
clean_df['Heart_Rate'].fillna(
    clean_df['Heart_Rate'].median(),
    inplace=True
)

In [ ]:
# Check remaining invalid heart rate values
invalid_hr = clean_df[
    (clean_df['Heart_Rate'] < 40) |
    (clean_df['Heart_Rate'] > 200)
]

print("Invalid Heart Rate Records:", len(invalid_hr))

In [ ]:
clean_df['Heart_Rate'].isnull().sum()

The Blood Pressure column was examined to identify inconsistent formats, missing values, and invalid entries before cleaning.

In [ ]:
# Check blood pressure values
# Identify formatting issues and invalid entries
clean_df['Blood_Pressure'].value_counts().head(20)

The Blood Pressure column contained the same measurement represented in multiple formats. To improve consistency, all blood pressure values were standardized into a single format (systolic/diastolic).

Example:

120 over 80 → 120/80
120 - 80 → 120/80

In [ ]:
# Standardize blood pressure format
clean_df['Blood_Pressure'] = clean_df['Blood_Pressure'].replace(
    '120 over 80',
    '120/80'
)
clean_df['Blood_Pressure'] = clean_df['Blood_Pressure'].replace(
    '120 - 80',
    '120/80'
)

Now you should see:

120/80

with a much larger count.

In [ ]:
# Check blood pressure values after cleaning

clean_df['Blood_Pressure'].value_counts().head(20)

In [ ]:
# Check missing blood pressure values
clean_df['Blood_Pressure'].isnull().sum()

The Blood Pressure column contained 2,030 missing records. Since blood pressure values are stored as text in systolic/diastolic format, missing values were replaced with "Unknown" instead of using a numerical estimate.

In [ ]:
# Fill missing blood pressure values
clean_df['Blood_Pressure'].fillna('Unknown', inplace=True)

In [ ]:
# Check missing blood pressure values
clean_df['Blood_Pressure'].isnull().sum()

In [ ]:
# Check cholesterol values
clean_df['Cholesterol_Level'].value_counts(dropna=False)

The Cholesterol column contained both numeric and text values. To ensure consistency, text categories were converted into their equivalent numeric representations.
normal → 190
high   → 250

In [ ]:
# Standardize cholesterol values
clean_df['Cholesterol_Level'] = clean_df['Cholesterol_Level'].replace(
    'normal', 190
)

clean_df['Cholesterol_Level'] = clean_df['Cholesterol_Level'].replace(
    'high', 250
)

In [ ]:
# Convert cholesterol to numeric
clean_df['Cholesterol_Level'] = pd.to_numeric(
    clean_df['Cholesterol_Level'],
    errors='coerce'
)

Missing cholesterol values were replaced using the median cholesterol level to maintain dataset completeness while avoiding data loss.

In [ ]:
# Fill missing cholesterol values
clean_df['Cholesterol_Level'].fillna(
    clean_df['Cholesterol_Level'].median(),
    inplace=True
)

In [ ]:
# Check missing cholesterol values
clean_df['Cholesterol_Level'].isnull().sum()

In [ ]:
clean_df['Gender'].value_counts(dropna=False)

In [ ]:
# Standardize gender values

clean_df['Gender'] = clean_df['Gender'].replace({
    'M': 'Male',
    'male': 'Male',
    'F': 'Female',
    'FEMALE': 'Female'
})

In [ ]:
# Fill missing gender values
clean_df['Gender'].fillna('Unknown', inplace=True)

The Gender column contained multiple representations of the same category, such as "M" and "male" for Male, and "F" and "FEMALE" for Female. These values were standardized into consistent categories. Missing values were replaced with "Unknown" to maintain completeness of the dataset.

In [ ]:
# Check gender values after cleaning
clean_df['Gender'].value_counts(dropna=False)

In [ ]:
# Check city values
clean_df['City'].value_counts(dropna=False)

In [ ]:
# Standardize city names
# Fix spelling and formatting issues
clean_df['City'] = clean_df['City'].replace({
    'chicago': 'Chicago',
    'NY': 'New York',
    'newyork': 'New York',
    'Nwe Yrok': 'New York',
    'la': 'Los Angeles'
})

In [ ]:
# Fill missing city values

clean_df['City'].fillna('Unknown', inplace=True)

The City column contained spelling mistakes, abbreviations, inconsistent capitalization, and missing values. Similar city names such as NY, newyork, and Nwe Yrok were standardized into New York. Likewise, la was converted to Los Angeles and chicago was converted to Chicago. Missing values were replaced with Unknown to maintain consistency.

In [ ]:
# Check city values after cleaning
clean_df['City'].value_counts(dropna=False)

In [ ]:
# Check diabetic values
clean_df['Diabetic'].value_counts(dropna=False)

In [ ]:
# Standardize diabetic values
# Convert similar values into one format
# Standardize diabetic values

clean_df['Diabetic'] = clean_df['Diabetic'].replace({
    'y': 'Yes',
    'N': 'No',
    'no': 'No'
})

The Diabetic column contained inconsistent representations such as "y", "Yes", "N", and "no". These values were standardized into consistent categories: Yes and No. Missing values were replaced with Unknown to maintain dataset completeness and consistency.

In [ ]:
# Replace missing values with "Unknown"
clean_df['Diabetic'] = clean_df['Diabetic'].fillna('Unknown')

In [ ]:
# Check diabetic values after cleaning
clean_df['Diabetic'].value_counts(dropna=False)

In [ ]:
# Check smoker values
clean_df['Smoker'].value_counts(dropna=False)

In [ ]:
# Standardize smoker values
# Convert similar values into one format
# Standardize smoker values
clean_df['Smoker'] = clean_df['Smoker'].replace({
    'yes': 'Yes',
    'EX-smoker': 'Former'
})

In [ ]:
# Fill missing smoker values

clean_df['Smoker'].fillna('Unknown', inplace=True)

The Smoker column contained inconsistent representations such as "yes", "EX-smoker", and "Former". Similar values were standardized into common categories. Missing values were replaced with "Unknown" to maintain consistency and completeness of the dataset.

In [ ]:
# Check smoker values after cleaning
clean_df['Smoker'].value_counts(dropna=False)

In [ ]:
# Check medication values
# Identify missing and inconsistent values
clean_df['Medications'].value_counts(dropna=False)

Standardize medication names
Convert similar formats into one format
Problem Identified
The same medications are written in different formats:
aspirin,metformin Aspirin; Metformin

Different separators:

comma (,) semicolon (;)

Different capitalization:

aspirin Aspirin

And missing values.

In [ ]:
# Standardize medication values
clean_df['Medications'] = clean_df['Medications'].replace({
    'aspirin,metformin': 'Aspirin, Metformin',
    'Aspirin; Metformin': 'Aspirin, Metformin',
    'Lisinopril / Metformin': 'Lisinopril, Metformin'
})

In [ ]:
# Fill missing medication values
clean_df['Medications'].fillna('Unknown', inplace=True)

The Medications column contained inconsistent formatting, including different separators and capitalization styles. Medication names were standardized into a consistent format. Missing values were replaced with "Unknown" to maintain dataset completeness.

In [ ]:
# Check medication values after cleaning
clean_df['Medications'].value_counts(dropna=False)

In [ ]:
# Check follow up values
clean_df['Follow_Up'].value_counts(dropna=False)

The Follow_Up column contained inconsistent representations such as "two weeks" and "14". These values were standardized into a common format. Missing values were replaced with "Unknown" to maintain consistency.

In [ ]:
# Standardize follow up values
clean_df['Follow_Up'] = clean_df['Follow_Up'].replace({
    'two weeks': '14'
})

In [ ]:
# Fill missing follow up values
clean_df['Follow_Up'].fillna('Unknown', inplace=True)

In [ ]:
# Check follow up values after cleaning
clean_df['Follow_Up'].value_counts(dropna=False)

In [ ]:
clean_df['Diagnosis_Code'].value_counts(dropna=False)

In [ ]:
clean_df['Has_Disease'].value_counts(dropna=False)

The Diagnosis_Code column contained missing values and inconsistent representations of unknown diagnoses. Missing values were replaced with "Unknown" and text formatting was standardized to improve consistency.

Current values:

B20
unknown
NaN
I10
E11.9
A00

Problem:

Only issue is:

unknown
NaN

Both represent unavailable information.

In [ ]:
# Fill missing diagnosis codes

clean_df['Diagnosis_Code'].fillna(
    'Unknown',
    inplace=True
)

In [ ]:
# Standardize diagnosis code values

clean_df['Diagnosis_Code'] = clean_df['Diagnosis_Code'].replace(
    'unknown',
    'Unknown'
)

In [ ]:
# Handle missing diagnosis codes
# Standardize diagnosis code values
# Check diagnosis code values after cleaning

clean_df['Diagnosis_Code'].value_counts(dropna=False)

In [ ]:
clean_df['Last_Visit_Date'].head(20)

First, I checked whether the Last_Visit_Date column contained any missing values. The result showed 0 missing values, which means every patient record has a date available.

In [ ]:
# Check missing values in date column
print("Missing Dates:",
      clean_df['Last_Visit_Date'].isnull().sum())

The date column contained multiple date formats. Some dates were written as text, some in YYYY-MM-DD format, some in DD/MM/YYYY format, and some as YYYYMMDD. This inconsistency can create problems during analysis.

In [ ]:
# View sample date values
clean_df['Last_Visit_Date'].head(10)

I converted all date formats into a standard datetime format using pandas. The format='mixed' option allows pandas to recognize different date formats automatically.

In [ ]:
# Convert mixed date formats into datetime format

clean_df['Last_Visit_Date'] = pd.to_datetime(
    clean_df['Last_Visit_Date'],
    format='mixed',
    errors='coerce'
)

After conversion, I checked whether any dates could not be interpreted. Invalid dates become NaT (Not a Time). The result showed 0 invalid dates, which means all dates were successfully converted.

In [ ]:
# Check invalid dates after conversion
print("Invalid Dates:",
      clean_df['Last_Visit_Date'].isnull().sum())

The Last_Visit_Date column contained multiple date formats such as "Apr 15 2021", "2021-07-10", "09/03/2024", and "20241113". I first checked for missing values and found none. Then I converted all dates into a standard datetime format using pandas mixed-format parsing. Finally, I verified that there were no invalid dates after conversion. The column is now fully standardized and ready for analysis.

In [ ]:
# Display converted dates
clean_df['Last_Visit_Date'].head(10)

In [ ]:
# Check notes values

clean_df['Notes'].value_counts(dropna=False)

In [ ]:
# Fill missing notes
clean_df['Notes'].fillna(
    'No Notes',
    inplace=True
)

The Notes column contained useful patient observations and comments. Since these notes may provide additional medical context, the column was retained. Missing values were replaced with "No Notes" to maintain consistency and completeness.

In [ ]:
# Check notes values after cleaning

clean_df['Notes'].value_counts(dropna=False)

In [ ]:
clean_df['Patient_ID'].isnull().sum()

In [ ]:
clean_df['Patient_ID'].head(20)

Patient_ID is a unique identifier used to distinguish records. Since it is not used for analysis or prediction, missing IDs were replaced with "Unknown_ID" to maintain dataset completeness.

In [ ]:
# Fill missing patient IDs
clean_df['Patient_ID'].fillna(
    'Unknown_ID',
    inplace=True
)

In [ ]:
clean_df['Has_Disease'].value_counts(dropna=False)

In [ ]:
# Check missing values in all columns
clean_df.isnull().sum()

In [ ]:
#considering Has_Disease column as target
clean_df['Has_Disease'].value_counts(dropna=False)

In [ ]:
# Standardize target values
clean_df['Has_Disease'] = clean_df['Has_Disease'].replace(
    'unknown',
    'Unknown'
)

In [ ]:
clean_df['Has_Disease'].value_counts(dropna=False)

In [ ]:
clean_df['Has_Disease'] = clean_df['Has_Disease'].fillna('Unknown')

In [ ]:
# View sample names
clean_df['Name'].head(20)

In [ ]:
# Standardize name format
# Standardize name capitalization(we use .str.title() function in pandas)
clean_df['Name'] = clean_df['Name'].str.title()

In [ ]:
# Add space after comma in names

clean_df['Name'] = clean_df['Name'].str.replace(
    ',',
    ', ',
    regex=False
)

The Name column contained formatting inconsistencies such as uppercase names, lowercase names, and missing spaces after commas. The names were standardized using title case formatting and consistent punctuation spacing to improve readability and data quality.

In [ ]:
clean_df['Name'].head(20)

In [ ]:
clean_df['BMI'].dtype

# Create BMI categories based on BMI ranges
# Underweight : BMI < 18.5
# Normal      : 18.5 - 24.9
# Overweight  : 25 - 29.9
# Obese       : BMI >= 30

In [ ]:
# Create BMI categories based on standard BMI ranges
clean_df['BMI_Category'] = pd.cut(
    clean_df['BMI'],
    bins=[0, 18.5, 25, 30, 100],
    labels=[
        'Underweight',
        'Normal',
        'Overweight',
        'Obese'
    ]
)

In [ ]:
# Display number of patients in each BMI category
# Check BMI category distribution

clean_df['BMI_Category'].value_counts()

In [ ]:
# Check dataset dimensions after adding BMI category

clean_df.shape

In [ ]:
clean_df[['Age','BMI','Heart_Rate']].describe()

In [ ]:
clean_df['Age'].dtype

In [ ]:
clean_df['Age'] = pd.to_numeric(clean_df['Age'], errors='coerce')

In [ ]:
clean_df['Age'].dtype

In [ ]:
# Detect age outliers using IQR
Q1 = clean_df['Age'].quantile(0.25)
Q3 = clean_df['Age'].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

age_outliers = clean_df[
    (clean_df['Age'] < lower_bound) |
    (clean_df['Age'] > upper_bound)
]

print("Age Outliers:", len(age_outliers))

The IQR method detected 2532 age outliers. However, most records in the dataset have an age value of 20, causing the interquartile range (IQR) to become zero. As a result, all ages different from 20 were identified as outliers. This is a limitation of the IQR method for highly concentrated data.

In [ ]:
import matplotlib.pyplot as plt

plt.boxplot(clean_df['Age'])

plt.title('Age Box Plot')
plt.ylabel('Age')
plt.show()

In [ ]:
clean_df['BMI'].dtype

In [ ]:
clean_df['BMI'] = clean_df['BMI'].astype(str)
clean_df['BMI'] = clean_df['BMI'].str.extract(r'(\d+\.?\d*)')[0]
clean_df['BMI'] = pd.to_numeric(clean_df['BMI'], errors='coerce')

In [ ]:
clean_df['BMI'].dtype

In [ ]:
# Detect BMI Outliers using IQR

Q1 = clean_df['BMI'].quantile(0.25)
Q3 = clean_df['BMI'].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

bmi_outliers = clean_df[
    (clean_df['BMI'] < lower_bound) |
    (clean_df['BMI'] > upper_bound)
]

print("BMI Outliers:", len(bmi_outliers))

The IQR method detected 2253 BMI outliers. However, these values fall within medically valid BMI ranges such as Underweight, Overweight, and Obese categories. Therefore, they were retained rather than removed to preserve meaningful health information.

In [ ]:
import matplotlib.pyplot as plt

plt.boxplot(clean_df['BMI'])

plt.title('BMI Box Plot')
plt.ylabel('BMI')
plt.show()

In [ ]:
# Detect heart rate outliers using IQR

Q1 = clean_df['Heart_Rate'].quantile(0.25)
Q3 = clean_df['Heart_Rate'].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

heart_outliers = clean_df[
    (clean_df['Heart_Rate'] < lower_bound) |
    (clean_df['Heart_Rate'] > upper_bound)
]

print("Heart Rate Outliers:", len(heart_outliers))

In [ ]:
# Visualize heart rate outliers using box plot

plt.boxplot(clean_df['Heart_Rate'])
plt.title('Heart Rate Box Plot')
plt.ylabel('Heart Rate')
plt.show()

In [ ]:
# Display dataset information after cleaning

clean_df.info()

In [ ]:
# Check missing values in all columns

clean_df.isnull().sum()

Why is Has_Disease still having missing values?
Answer:
Has_Disease is the target variable. Since the actual disease status is unknown for these records, assigning 0 or 1 would introduce incorrect information. Therefore, the missing values were retained.

In [ ]:
# Check duplicate records

clean_df.duplicated().sum()

In [ ]:
# Display dataset dimensions

clean_df.shape

In [ ]:
# Display all columns after cleaning

print(clean_df.columns.tolist())

#Concept of Encoding
What is Encoding?

Machine Learning models cannot understand:

Male
Female
Chicago
Boston
Yes
No

They understand only numbers.

So we convert text into numbers.

This process is called Encoding.

One-Hot Encoding:

Example:

Before:

Gender
Male
Female
Unknown

After:

Gender_Male	Gender_Female	Gender_Unknown
1	0	0
0	1	0
0	0	1

In [ ]:
# Create encoded dataset
# Apply one-hot encoding to categorical columns
encoded_df = pd.get_dummies(
    clean_df,
    columns=[
        'Gender',
        'City',
        'Diabetic',
        'Smoker',
        'BMI_Category'
    ]
)

In [ ]:
#After one-hot encoding, categorical columns are converted into multiple binary columns. Therefore, the number of columns increases.
encoded_df.shape

In [ ]:
#This step verifies the newly created encoded features.
encoded_df.columns.tolist()

Correlation measures the relationship between numerical variables. Values range from -1 to +1.

+1 → Strong positive relationship
-1 → Strong negative relationship
0 → No relationship

In [ ]:
# Calculate correlation between numerical columns
corr_matrix = clean_df[
    ['Age', 'BMI', 'Heart_Rate', 'Cholesterol_Level']
].corr()

corr_matrix

Correlation Results
Variables	Correlation
Age vs BMI	0.013
Age vs Heart Rate	0.014
Age vs Cholesterol	0.008
BMI vs Heart Rate	-0.005
BMI vs Cholesterol	-0.003
Heart Rate vs Cholesterol	0.015
# Calculate correlation among numerical features
Explanation

Correlation analysis was performed to identify relationships between numerical variables such as Age, BMI, Heart Rate, and Cholesterol Level.

The correlation values are very close to 0, indicating that there is no strong linear relationship between these variables in the dataset.

Therefore, the features are largely independent of each other.

What did we observe?
Answer:
Most correlation values were close to zero, indicating weak or no significant relationship among the numerical features.


In [ ]:
# Visualize correlation matrix

plt.figure(figsize=(6,4))

sns.heatmap(
    corr_matrix,
    annot=True,
    cmap='coolwarm'
)

plt.title('Correlation Heatmap')
plt.show()

This chart shows the number of patients in each BMI category. Most patients belong to the Normal BMI category.

In [ ]:
# Visualize patient distribution across BMI categories

clean_df['BMI_Category'].value_counts().plot(
    kind='bar'
)

plt.title('BMI Category Distribution')
plt.xlabel('BMI Category')
plt.ylabel('Number of Patients')
plt.show()

This pie chart shows the percentage distribution of Male, Female, and Unknown gender categories.

In [ ]:
# Visualize gender distribution

clean_df['Gender'].value_counts().plot(
    kind='pie',
    autopct='%1.1f%%'
)

plt.title('Gender Distribution')
plt.ylabel('')
plt.show()

Shows the number of patients from each city.

In [ ]:
import matplotlib.pyplot as plt

plt.figure(figsize=(8,6))
plt.scatter(clean_df['Age'], clean_df['BMI'], alpha=0.5)

plt.title('Age vs BMI')
plt.xlabel('Age')
plt.ylabel('BMI')

plt.show()

In [ ]:
# Visualize patient distribution across cities

clean_df['City'].value_counts().plot(
    kind='bar'
)

plt.title('Patient Distribution by City')
plt.xlabel('City')
plt.ylabel('Number of Patients')
plt.show()

Shows how many patients are diabetic, non-diabetic, or unknown.

In [ ]:
# Visualize diabetic status

clean_df['Diabetic'].value_counts().plot(
    kind='bar'
)

plt.title('Diabetic Status Distribution')
plt.xlabel('Diabetic Status')
plt.ylabel('Count')
plt.show()

In [ ]:
# Visualize smoker status

clean_df['Smoker'].value_counts().plot(
    kind='bar'
)

plt.title('Smoking Status Distribution')
plt.xlabel('Smoking Status')
plt.ylabel('Count')
plt.show()

In [ ]:
# Visualize target variable distribution
clean_df['Has_Disease'].value_counts(
    dropna=False
).plot(
    kind='bar'
)

plt.title('Disease Status Distribution')
plt.xlabel('Disease Status')
plt.ylabel('Count')
plt.show()

In [ ]:
# Visualize diagnosis code distribution

clean_df['Diagnosis_Code'].value_counts().plot(
    kind='bar'
)

plt.title('Diagnosis Code Distribution')
plt.xlabel('Diagnosis Code')
plt.ylabel('Count')
plt.show()

In [ ]:
clean_df['Age'].value_counts().head(10)

In [ ]:
# Visualize age distribution

plt.hist(
    clean_df['Age'],
    bins=20
)

plt.title('Age Distribution')
plt.xlabel('Age')
plt.ylabel('Frequency')
plt.show()

In [ ]:
# Visualize BMI distribution

plt.hist(
    clean_df['BMI'],
    bins=20
)

plt.title('BMI Distribution')
plt.xlabel('BMI')
plt.ylabel('Frequency')
plt.show()

In [ ]:
# Visualize heart rate distribution

plt.hist(
    clean_df['Heart_Rate'],
    bins=20
)

plt.title('Heart Rate Distribution')
plt.xlabel('Heart Rate')
plt.ylabel('Frequency')
plt.show()

In [ ]:
# Visualize age outliers

plt.boxplot(
    clean_df['Age']
)

plt.title('Age Box Plot')
plt.ylabel('Age')
plt.show()

In [ ]:
clean_df['Age'].value_counts().head(10)

In [ ]:
# Visualize BMI outliers

plt.boxplot(
    clean_df['BMI']
)

plt.title('BMI Box Plot')
plt.ylabel('BMI')
plt.show()

In [ ]:
# Visualize heart rate outliers

plt.boxplot(
    clean_df['Heart_Rate']
)

plt.title('Heart Rate Box Plot')
plt.ylabel('Heart Rate')
plt.show()

In [ ]:
# Visualize correlation between numerical features

corr_matrix = clean_df[
    ['Age','BMI','Heart_Rate','Cholesterol_Level']
].corr()

sns.heatmap(
    corr_matrix,
    annot=True
)

plt.title('Correlation Heatmap')
plt.show()

In [ ]:
import matplotlib.pyplot as plt

# Convert labels for better readability (if not already done)
clean_df['Has_Disease'] = clean_df['Has_Disease'].replace({
    0: 'No Disease',
    1: 'Has Disease',
    '0': 'No Disease',
    '1': 'Has Disease'
})

plt.figure(figsize=(8,6))

clean_df.boxplot(column='BMI', by='Has_Disease')

plt.title('BMI vs Disease Status')
plt.suptitle('')  # Removes the automatic title
plt.xlabel('Disease Status')
plt.ylabel('BMI')

plt.show()

In [ ]:
# Replace disease labels for better visualization
clean_df['Has_Disease'] = clean_df['Has_Disease'].replace({
    0: 'No Disease',
    1: 'Has Disease',
    '0': 'No Disease',
    '1': 'Has Disease'
})

# Create grouped bar chart for Diabetic vs Has_Disease
pd.crosstab(
    clean_df['Diabetic'],
    clean_df['Has_Disease']
).plot(
    kind='bar',
    figsize=(8,6)
)

plt.title('Diabetic Status vs Disease Status')
plt.xlabel('Diabetic Status')
plt.ylabel('Count')
plt.legend(title='Disease Status')
plt.show()

In [ ]:
# Replace disease labels for better visualization
clean_df['Has_Disease'] = clean_df['Has_Disease'].replace({
    0: 'No Disease',
    1: 'Has Disease',
    '0': 'No Disease',
    '1': 'Has Disease'
})

# Compare BMI category with disease status
pd.crosstab(
    clean_df['BMI_Category'],
    clean_df['Has_Disease']
).plot(
    kind='bar'
)

plt.title('BMI Category vs Disease Status')
plt.xlabel('BMI Category')
plt.ylabel('Count')
plt.legend(title='Disease Status')
plt.show()

In [ ]:
# Compare gender and disease status

pd.crosstab(
    clean_df['Gender'],
    clean_df['Has_Disease']
).plot(
    kind='bar'
)

plt.title('Gender vs Disease Status')
plt.xlabel('Gender')
plt.ylabel('Count')
plt.show()

In [ ]:
# Move BMI_Category next to BMI

cols = list(clean_df.columns)

cols.insert(
    cols.index('BMI') + 1,
    cols.pop(cols.index('BMI_Category'))
)

clean_df = clean_df[cols]

In [ ]:
clean_df.columns.tolist()

In [ ]:
# Save cleaned dataset

clean_df.to_csv(
    'Cleaned_Healthcare_Dataset.csv',
    index=False
)

In [ ]:
import os
os.listdir()

In [ ]:
from google.colab import files

files.download('Cleaned_Healthcare_Dataset.csv')